File: 437-path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        
        def pathSum_a(root, sum):
            if root == None: return 0
            res = 0
            if root.val == sum: res += 1
            res += pathSum_a(root.left, sum-root.val)
            res += pathSum_a(root.right, sum-root.val)
            return res
        
        if root == None: return 0
        return self.pathSum(root.left, targetSum) + pathSum_a(root, targetSum) + self.pathSum(root.right, targetSum)

File: 437-path-sum-iii/path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int pathSum_a(TreeNode* root, long long sum) {
        if (root == nullptr) return 0;
        int res = 0;
        if (root->val == sum) res++;
        res += pathSum_a(root->left, sum-root->val);
        res += pathSum_a(root->right, sum-root->val);
        return res;
    }
public:
    int pathSum(TreeNode* root, int targetSum) {
       if (root == nullptr) return 0;
       return pathSum(root->left, targetSum) + pathSum_a(root, targetSum) + pathSum(root->right, targetSum);
    }
};